### ISSUE TITLE
Login/user management

### ISSUE THREAD

**Issue by interrogator** [interrogator/buzzword#4]
The biggest feature still required is the ability to login and see your stored datasets. Backend is nearly prepared for this, but buzzword will need a db, captcha, etc

**Comment by interrogator** [comment]
Fairly complete example: https://hackersandslackers.com/authenticating-users-with-flask-login/

Feature-rich is important, because once logged in users will need to be able to set preferences and so on.

There are dash-specific logins, but I am worried about lack of development on them.

https://github.com/RafaelMiquelino/dash-flask-login

**Comment by interrogator** [comment]
Note, above links are not needed if development continues in django.

User management will be easier in django. OAuth would be good. Most users will be academic, so supporting university credentials would be good.

**Comment by interrogator** [comment]
@l0rb can you quickly tell me where you're at on this, what's done and to do and so on? Just some quick bullet points

**Comment by l0rb** [comment]
- [x] users can register
- [x] login
- [x] logout
- [ ] require email for signup
- [ ] change password
- [ ] password forgotten

these are the basic features that I believe are the bare minimum we need.

**Comment by interrogator** [comment]
Yes, those are obvious and needed. Aside from those bits, we have been speaking already about permissions. I think best approach is to make a `Permissions` model, and link it to User model by foreign key. So: `user.permissions.can_upload`.

`Permissions` can model the set of permission types. I believe new permission types could appear at any point. If a uni signs up, they might want certain features and not need others. In this case, we can create a new entry in `Permissions` that suits them, and users coming from their domain will automatically be assigned the right 'abilities'

**Comment by l0rb** [comment]
Yes this is pretty much how it works. the name of the relation (foreign key) is `user_permissions` though, and there is also `user.groups`. A user can be given a permission with `user.user_permissions.add(permission)`. checking if a user has a permission has a shortcut: `user.has_perm(permission)`
